pkg/fqdn/dnspoller.go

```diff
@@ -0,0 +1,373 @@
+// Copyright 2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fqdn
+
+import (
+	"bytes"
+	"net"
+	"sort"
+	"time"
+
+	"github.com/cilium/cilium/pkg/controller"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/policy/api"
+
+	"github.com/sirupsen/logrus"
+)
+
+// generatedLabelNameUUID is the label key for policy rules that contain a
+// ToFQDN section and need to be updated
+const generatedLabelNameUUID = "ToFQDN-UUID"
+
+// uuidLabelSearchKey is an *extended* label key. This is because .Has
+// expects the source:key delimiter to be the labels.PathDelimiter
+var uuidLabelSearchKey = generateUUIDLabel().GetExtendedKey()
+
+// StartDNSPoller spawns a singleton DNS polling controller. The controller
+// will, periodically, run a DNS lookup for each ToFQDN target DNS name
+// inserted with StartPollForDNSName.
+// Note: Repeated calls will replace earlier instances of the controller.
+func StartDNSPoller(poller *DNSPoller) {
+	log.Debug("Starting DNS poller for ToFQDN rules")
+	controller.NewManager().UpdateController("dns-poller", controller.ControllerParams{
+		RunInterval: 5 * time.Second,
+		DoFunc:      poller.LookupUpdateDNS,
+		StopFunc: func() error {
+			log.Debug("Stopping DNS poller for ToFQDN rules")
+			return nil
+		},
+	})
+}
+
+// DNSPoller periodically runs lookups for registered DNS names. It will emit
+// regenerated policy rules when the IPs change. CNAMEs (and DNAMEs) are not
+// handled directly, but will depend on the resolver's behavior.
+// DNSPollerConfig can be opitonally used to set how the DNS lookups are
+// executed (via LookupDNSNames) and how generated policy rules are handled
+// (via AddGeneratedRules).
+type DNSPoller struct {
+	lock.Mutex // this guards both maps and their contents
+
+	// config is a copy from when this instance was initialized.
+	// It is read-only once set
+	config DNSPollerConfig
+
+	// IPs maps dnsNames as strings to the most recent IPs seen for them
+	// Note: The IP slice is sorted. on insert in updateIPsForName, and should
+	// not be reshuffled.
+	IPs map[string][]net.IP
+
+	// sourceRule maps dnsNames to a map of rule UUID to rules that
+	// depend on that dnsName.
+	// The map-in-map is used to simplify rule removal.
+	sourceRules map[string]map[string]*api.Rule
```

Comment:
I'm having a little trouble getting my head around the management of `IPs` and `sourceRules` here. Could the `sourceRules` be considered the "desired" resolvable FQDNs, then `IPs` be considered the "realized" state of the mapping from that FQDN to some set of IPs?

One reason I ask is that it seems like in `daemon.policyAdd()`, we have the logic to start trying to poll these FQDNs. Let's say this is triggered by a user adding a policy and it everything works fine. Then some time later, the controller polls and resolves some new IPs, and it triggers a callback to `AddGeneratedRules`, which then calls back into `daemon.PolicyAdd()` again, but this time maybe there's some problem. Like a connectivity issue to kvstore or something, I'm not exactly sure. So then the `PolicyAdd()` call actually fails and calls `StopPollForDNSName()`. Does this callback now disable the FQDN resolver for this policy?

Reply:
Is it as simple as making sure that the rules that are passed to `AddGeneratedRules()` are guaranteed to be only generated, then `StartPollForDNSName()` and  `StopPollForDNSName()` ignores generated rules?

Reply:
I augment an existing policy thing* in-place, basically, and the AddGeneratedRules adds them with replace. I could go back to an earlier scheme were I made wholly new rules that live side-by-side with the user inserted ones.
Unfortunately, StopPollForDNSName currently only does something when a rule has been marked but even if it didn't we would need to decrement the refcount (or remove the rule pointer) in some remove function. That function has to be the same one unless we change policyAdd's logic somewhat, and pass more data back and forth.

* I honestly still don't know what a rule actually is vs the various directives like ToEndpoints and ToPorts so I'm probably making less sense.